perception/autoware_tensorrt_bevdet/src/bevdet.cpp

```diff
@@ -0,0 +1,717 @@
+                                cams2ego_trans_[i].translation();
+
+          // voxelization
+          *(frustum + offset) -= Eigen::Vector3f(x_start_, y_start_, z_start_);
+          (frustum + offset)->x() = static_cast<int>((frustum + offset)->x() / x_step_);
+          (frustum + offset)->y() = static_cast<int>((frustum + offset)->y() / y_step_);
+          (frustum + offset)->z() = static_cast<int>((frustum + offset)->z() / z_step_);
+        }
+      }
+    }
+  }
+
+  int * _ranks_depth = new int[num_points];
+  int * _ranks_feat = new int[num_points];
+
+  for (int i = 0; i < num_points; i++) {
+    _ranks_depth[i] = i;
+  }
+  for (int i = 0; i < n_img_; i++) {
+    for (int d_ = 0; d_ < depth_num_; d_++) {
+      for (int u = 0; u < feat_h_ * feat_w_; u++) {
+        int offset = i * (depth_num_ * feat_h_ * feat_w_) + d_ * (feat_h_ * feat_w_) + u;
+        _ranks_feat[offset] = i * feat_h_ * feat_w_ + u;
+      }
+    }
+  }
+
+  std::vector<int> kept;
+  for (int i = 0; i < num_points; i++) {
+    if (
+      static_cast<int>((frustum + i)->x()) >= 0 &&
+      static_cast<int>((frustum + i)->x()) < xgrid_num_ &&
+      static_cast<int>((frustum + i)->y()) >= 0 &&
+      static_cast<int>((frustum + i)->y()) < ygrid_num_ &&
+      static_cast<int>((frustum + i)->z()) >= 0 &&
+      static_cast<int>((frustum + i)->z()) < zgrid_num_) {
+      kept.push_back(i);
+    }
+  }
+
+  valid_feat_num_ = kept.size();
+  int * ranks_depth_host = new int[valid_feat_num_];
+  int * ranks_feat_host = new int[valid_feat_num_];
+  int * ranks_bev_host = new int[valid_feat_num_];
+  int * order = new int[valid_feat_num_];
+
+  for (int i = 0; i < valid_feat_num_; i++) {
+    Eigen::Vector3f & p = frustum[kept[i]];
+    ranks_bev_host[i] = static_cast<int>(p.z()) * xgrid_num_ * ygrid_num_ +
+                        static_cast<int>(p.y()) * xgrid_num_ + static_cast<int>(p.x());
+    order[i] = i;
+  }
+
+  thrust::sort_by_key(ranks_bev_host, ranks_bev_host + valid_feat_num_, order);
+  for (int i = 0; i < valid_feat_num_; i++) {
+    ranks_depth_host[i] = _ranks_depth[kept[order[i]]];
+    ranks_feat_host[i] = _ranks_feat[kept[order[i]]];
+  }
+
+  delete[] _ranks_depth;
+  delete[] _ranks_feat;
+  delete[] frustum;
+  delete[] order;
+
+  std::vector<int> interval_starts_host;
+  std::vector<int> interval_lengths_host;
+
+  interval_starts_host.push_back(0);
+  int len = 1;
+  for (int i = 1; i < valid_feat_num_; i++) {
+    if (ranks_bev_host[i] != ranks_bev_host[i - 1]) {
+      interval_starts_host.push_back(i);
+      interval_lengths_host.push_back(len);
+      len = 1;
+    } else {
+      len++;
+    }
+  }
+
+  interval_lengths_host.push_back(len);
+  unique_bev_num_ = interval_lengths_host.size();
+
+  int * interval_starts_host_ptr = new int[interval_starts_host.size()];
+  int * interval_lengths_host_ptr = new int[interval_lengths_host.size()];
```

Comment:
Memory Leak: Memory allocated here is not freed.  `std::vector` should be preferred for `interval_starts_host` and `interval_lengths_host`.